[DOC/CI] Make installation documentation and CI use the same commands

pjones opened the pull request:
## Description

Prevent the documentation from becoming stale by using the same dependency installation commands as CI.

----

I have extracted the dependency installation commands from the CI workflow and put them in their own shell scripts.  This allows the CI workflow *and* the installation documentation to use the same installation commands.  Hopefully this means that the documentation is always up to date with respect to the source code.

Furthermore I'm opening this as a draft because I'd like to start a discussion about the instructions for using `contrib`.  Since we do not use `contrib` in CI for macOS and Linux, we can't be sure that they can successfully link with OpenMS on those platforms.  I would therefore like to remove the instructions for using `contrib` from the macOS and Linux installation instructions and focus on using the same commands we use in CI.  (I also think we should move away from using `contrib` on Windows and use a dedicated package manager like we do on the other platforms.  But that's for another day.)

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **Documentation**
  - Added and reorganized installation guides for Linux and macOS, providing streamlined, modular instructions and improved clarity.
  - Introduced new documentation for building dependencies from source and for testing OpenMS installations.
  - Enhanced CMake parameter documentation with additional usage tips.

- **Chores**
  - Centralized and automated dependency installation for Ubuntu and macOS using new scripts, improving maintainability and consistency in setup workflows.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

pjones committed:
[DOC/CI] Make installation documentation and CI use the same commands

Prevent the documentation from becoming stale by using the same
dependency installation commands as CI.

timosachsenberg commented:
I like the scripts.

When we discussed getting rid of contrib typically those two arguments were brought up:
- If a package is not available on a certain Linux distro or architecture, then users can't build it. This might be relevant for e.g. instrument vendors or some hpc setups.
- Same if we want to add a new dependency

That being said. I am much in favor of reducing the role of contrib.
If maintainable, I would preferhaving two ways to build OpenMS: with 
1. system packages for binaries
2. vcpkg for full source builds  (currently not really tested or used)

pjones commented:
I do think that we should continue to allow building with `contrib`, at least until the vcpkg stuff is ready.  All I want to do is to update the installation instructions.  However, we could include a blurb at the end of the docs saying how to build using `contrib` for those who don't have access to the packages.

timosachsenberg commented:
got ya. Yeah totally makes sense.

pjones committed:
[DOC] Rewrite build instructions for macOS and Linux

Streamline the build instructions and demote instructions for using
contrib.

pjones commented:
@timosachsenberg Ready for final review.

jpfeuffer commented on doc/doxygen/install/install-macs.doxygen:
```diff
@@ -4,255 +4,116 @@
-  <ul>
-   <li><a href="http://www.doxygen.org/index.html">Doxygen</a> (recommended: >= 1.8.13, required: > 1.7.1, some might be buggy in between) is used to generate the %OpenMS documentation.</li>
-  </ul>
+    A minimal installation of Apple Clang is achieved by downloading
+    Apple's "Command Line Tools".  However we suggest using a full
+    Xcode installation from the Mac App Store.  %OpenMS requires
+    version 4.2 or greater.  More details can be found on the [Apple
+    Developer Site][xcode].
 
```
Valid point 😅

timosachsenberg replied:
yes please :)

pjones replied:
Updated.

jpfeuffer commented on doc/doxygen/install/install-macs.doxygen:
````diff
@@ -4,255 +4,116 @@
+      -DBOOST_USE_STATIC=OFF \
+      "$OPENMS_DIR/OpenMS"
+    make -j8
+    ```
````
Good point

jpfeuffer commented on tools/ci/deps-macos.sh:
```diff
@@ -0,0 +1,38 @@
+brew update
+
+# Required dependencies:
+brew install --force --overwrite \
+  python@3.13 \
+  ccache \
+  autoconf \
+  automake \
+  libtool \
+  ninja \
+  libomp \
```
True. Force overwrite is not something you want to suggest to users for their daily system.

pjones replied:
I've updated the script to remove these flags.  However I need to make sure this still works in CI so I've marked this PR as a draft.

pjones replied:
Fixed and verified that CI works without the `--force` flag.  It does require the `--overwrite` flag but I hid that from the documentation.

pjones commented on tools/ci/deps-ubuntu.sh:
```diff
@@ -0,0 +1,49 @@
+# Code between the following doxygen markers are included in the
+# public-facing OpenMS installation instructions.
+
+# [installation_documentation]
+# Add "universe" and update:
+sudo add-apt-repository universe
+sudo apt update
+
```
This isn't a problem.

pjones commented on doc/doxygen/install/install-linux.doxygen:
````diff
@@ -2,410 +2,140 @@
 
-  <br/>
-  <br/>
+    ```sh
+    mkdir -p "$OPENMS_DIR/openms_build"
+    cd "$OPENMS_DIR/openms_build"
+    cmake -DBOOST_USE_STATIC=OFF "$OPENMS_DIR/OpenMS"
+    make -j8
+    ```
````
I updated the instructions.

pjones committed:
[DOC] Update based on review feedback

jpfeuffer commented on tools/ci/deps-ubuntu.sh:
```diff
@@ -0,0 +1,49 @@
+  git \
+  automake \
+  ninja-build \
+  xvfb \
```
This is only needed if the computer does not have a monitor connected.

pjones replied:
Apparently this is a common enough use case that we need to support it.  We discussed this on Discord and people still build OpenMS on clusters and then remotely run the graphical applications via X11 forwarding via SSH.

pjones committed:
[DOCS/CI] We don't need to pin Python to version 3.13

This is probably an artifact for an older version of macOS.

pjones committed:
[DOC] Stick with Doxygen 1.9 Syntax

We still have to use Doxygen 1.9 on Ubuntu and it doesn't support
indented code fences.

pjones commented on tools/ci/deps-macos.sh:
```diff
@@ -0,0 +1,38 @@
+  doxygen \
+  ghostscript \
+  graphviz
+# [installation_documentation]
```
Removed `--force` and `--overwrite`.

pjones commented:
Ready for review again.

I implemented all suggested changes and verified that CI is building correctly.  The Windows build fails during the NSIS step because my branch name makes file names too long.  Otherwise everything is good.

pjones commented on doc/doxygen/install/install-linux.doxygen:
```diff
@@ -2,410 +2,155 @@
-  </table>
+  Dependencies can be installed using the following commands:
   
+  \snippet "../../tools/ci/deps-ubuntu.sh" installation_documentation
```
Wrong.  Doxygen is run with "$(pwd)" in a weird place.

pjones commented on doc/doxygen/install/install-linux.doxygen:
```diff
@@ -2,410 +2,155 @@
-  @note Since static versions of the boost library are on some systems (e.g. Ubuntu) compiled with different settings you have to set BOOST_USE_STATIC to OFF or compile it with our contrib.
   \code
-  cmake -DOPENMS_CONTRIB_LIBS="/PATH/TO/contrib-build" -DBOOST_USE_STATIC=OFF ../OpenMS
+  # register obs://build.opensuse.org/devel:libraries:c_c++
```
This is a known issue and these instructions are marked as unmaintained.

timosachsenberg committed:
Merge branch 'develop' into pjones/ci-install

timosachsenberg reviewed (approved):
nice improvement
